File: Spline.java
import java.text.*;
import java.util.*;

public class Spline// implements Interpolation
{
    int N; // Size of given data set.

    // given data
    public double[] x;   // 0 - N-1
    public double[] f;   // 0 - N-1

    double[] h;   // 0 - N-2
    double[] f1;  // 0 - N-2
    double[] f2;  // 1 - N-2

    public Spline(double[] _x, double[] _f, int _N)
    {
        this(_x, _f, _N, 0.0, 0.0);
    }
    public Spline(double[] _x, double[] _f, int _N, double boundL, double boundR)
    {
        double[][] _tmpA = new double[_N][2];
        for(int i=0; i<_N; i++)
            {
                _tmpA[i][0] = _x[i];
                _tmpA[i][1] = _f[i];
            }

        double[] _tmpX = new double[_N];
        double[] _tmpF = new double[_N];
        for(int i=0; i<_N; i++)
            {
                _tmpX[i] = _tmpA[i][0];
                _tmpF[i] = _tmpA[i][1];
            }
        setDataSet(_tmpX, _tmpF, _N, boundL, boundR);
    }

    public boolean setDataSet(double[] _x, double[] _f, int _N,
                              double boundL, double boundR)
    {
        N = _N;

        x = new double[N];
        f = new double[N];
        h = new double[N-1];
        f1 = new double[N-1];
        f2 = new double[N];

        for(int i=0; i<N; i++)
            {
                x[i] = _x[i];
                f[i] = _f[i];
            }

        for(int i=0; i<N-1; i++) h[i] = x[i+1] - x[i];

      
        _set_y2(boundL, boundR);
        _set_y1();

        return true;
    }

    boolean _set_y2(double boundL, double boundR)
    {
        double[][] H = new double[N-2][N-2];
        double[] y = new double[N-2];
        double[] b = new double[N-2];

        for(int i=0; i<N-2; i++)
            H[i][i] = 2 * ( h[i] + h[i+1]);
        for(int i=0; i<N-3; i++)
            H[i][i+1] = h[i+1];
        for(int i=1; i<N-2; i++)
            H[i][i-1] = h[i-1];

        for(int i=0; i<N-2; i++)
            b[i] = 
                6.0 * ( (f[i+2] - f[i+1])/h[i+1] - (f[i+1] - f[i])/h[i] );

        Gauss g = new Gauss(H, b);
        g.gauss();
        y = g.x;
      
        f2[0] = boundL;
        for(int i=1; i<N-1; i++) f2[i] = y[i-1];
        f2[N-1] = boundR;

        return true;
    }
  
    boolean _set_y1()
    {
        for(int i=0; i<N-1; i++)
            {
                f1[i] = (f[i+1] - f[i])/ h[i] - f2[i+1] * h[i] / 6.0 - f2[i] * h[i] / 3.0;
            }

        return true;
    }
  
    int _searchInterval(double _x)
    {
        if(_x < x[0]) return 0;

        for(int i=0; i<N-1; i++)
            {
                if( (x[i] <= _x) && (_x <= x[i+1]) ) return i;
            }
        return N-2;
    }

    double splinePoly(double _x, int i)
    {
        return
            f[i]
            + f1[i] * (_x - x[i])
            + f2[i] * Math.pow( (_x - x[i]), 2.0 )/2.0
            + (f2[i+1] - f2[i]) 
            * Math.pow( (_x - x[i]), 3.0)/(6.0 * h[i]);
    }
  
    double splinePolyDev1(double _x, int i)
    {
        return
            f1[i]
            + f2[i] * (_x - x[i])
            + (f2[i+1] - f2[i])
            * Math.pow( (_x - x[i]), 2.0)/(2.0 * h[i]);
    }

    public double f(double _x)
    {
        int i = _searchInterval(_x);
        return splinePoly(_x, i);
    }
    public double dev1(double _x)
    {
        int i = _searchInterval(_x);
        return splinePolyDev1(_x, i);
    }

    public double x(int i) { return x[i]; }
    public double f(int i) { return f[i]; }
    public int N() { return N; }

    public void stdout()
    {
        DecimalFormat form = new DecimalFormat("######.####");

        String str = new String();
        str += "x  h f f1 f2\n";
        for(int i=0; i<N; i++)
            {
                str += form.format(x[i]);

                if(i <= N-2)
                    str += " " + form.format(h[i]);
                else
                    str += " -----";

                str += " " + form.format(f[i]);

                if(i <= N-2)
                    str += " " + form.format(f1[i]);
                else
                    str += " -----";

                str += " " + form.format(f2[i]);

                str += "\n";
            }
        System.out.println(str);
        return;
    }
  
    public static void main(String[] argv)
    {
        int n = 11;

        double[] x = new double[n];
        double[] y = new double[n];

        for(int i=0; i<n; i++) {
            x[i] = 10.0/(n-1) * i - 5;
            y[i] = Math.exp(- x[i]*x[i]);
        }

        Spline lsp = new Spline(x, y, n);
        //Spline lsp = new Spline(x, y, n, 0.5, 0.5);

        int N = 100;
        for(int i=0; i<=N; i++) {
            double testx = 14.0 /N * i - 7;
            System.out.println(testx + ", "+ lsp.f(testx));
        }

    }
}

File: spline.js
//'use strict'
function Spline(aX, aF, aN) {
    var sp = {};
    sp.x = copy(aX);   // 0 - N-1
    sp.f = copy(aF);
    sp.n = aN;
    sp.left_b = 0.0;
    sp.right_b = 0.0;

    sp.h = new Matrix(aN, 1);   // 0 - N-2
    sp.f1 = new Matrix(aN, 1);  // 0 - N-2
    sp.f2 = new Matrix(aN, 1); // 1 - N-2

    for(var i=0; i < sp.n-1; i++) sp.h.set(i, 0, sp.x.at(i+1, 0) - sp.x.at(i, 0));

    var H = new Matrix(sp.n-2, sp.n-2);
    //var y = new Matrix(sp.n-2, 1);
    var b = new Matrix(sp.n-2, 1);

    for(var i=0; i<sp.n-2; i++)  H.set(i, i,  2 * ( sp.h.at(i,0) + sp.h.at(i+1, 0)));
    for(var i=0; i<sp.n-3; i++)  H.set(i, i+1,  sp.h.at(i+1, 0));
    for(var i=1; i<sp.n-2; i++)  H.set(i, i-1,  sp.h.at(i-1, 0));

    for(var i=0; i<sp.n-2; i++)
    b.set(i, 0, 6.0 * ( (sp.f.at(i+2, 0) - sp.f.at(i+1, 0))/sp.h.at(i+1, 0) - (sp.f.at(i+1,0) - sp.f.at(i,0))/sp.h.at(i, 0) ) );

    var g = makeGauss(H, b);
    g.solve();
    //copy_set(g.x, y);

    sp.f2.set(0, 0, sp.left_b);
    for(var i=1; i<sp.n-1; i++) sp.f2.set(i, 0, g.x.at(i-1, 0));
    sp.f2.set(sp.n-1, 0, sp.right_b);

    //sp.f2.set(0, 0, (sp.f2.at(1, 0)+sp.f2.at(sp.n-2, 0))/2.0);
    //sp.f2.set(sp.n-1, 0, sp.f2.at(0,0));

    for(var i=0; i<sp.n-1; i++) {
        sp.f1.set(i, 0,   (sp.f.at(i+1, 0) - sp.f.at(i, 0))/ sp.h.at(i, 0) - sp.f2.at(i+1, 0) * sp.h.at(i, 0) / 6.0 - sp.f2.at(i, 0) * sp.h.at(i, 0) / 3.0 );
    }

    function _searchInterval(_x) {
        if(_x < sp.x.at(0, 0)) return 0;

        for(var i=0; i<sp.n-1; i++) {
            if( (sp.x.at(i, 0) <= _x) && (_x <= sp.x.at(i+1, 0)) ) {
                return i;
            }
        }
        return sp.n-2;
    }

    sp.lval = function(_x) {
        var i = _searchInterval(_x);
        return (sp.f.at(i+1, 0) - sp.f.at(i, 0))/(sp.x.at(i+1, 0)-sp.x.at(i,0)) * (_x-sp.x.at(i, 0)) + sp.f.at(i, 0);
    };

    sp.val = function(_x) {
        var i = _searchInterval(_x);
        return sp.f.at(i, 0)
        + sp.f1.at(i, 0) * (_x - sp.x.at(i, 0))
        + sp.f2.at(i, 0) * Math.pow( (_x - sp.x.at(i, 0)), 2.0 )/2.0
        + (sp.f2.at(i+1, 0) - sp.f2.at(i, 0))
        * Math.pow( (_x - sp.x.at(i, 0)), 3.0)/(6.0 * sp.h.at(i, 0));
    };

    return sp;
}


var n = 11;

var x = new Matrix(n, 1);
var y = new Matrix(n, 1);

function test_f(_x) {
    //return Math.sin(- _x * _x);
    return Math.sin(- _x);
}
for(var i=0; i<n; i++) {
    x.set(i, 0,  10.0/(n-1) * (2*i*i) - 5.0);
    y.set(i, 0,  test_f(x.at(i, 0)));
}


var lsp = Spline(x, y, n);
//Spline lsp = new Spline(x, y, n, 0.5, 0.5);

var N = 100;
for(var i=0; i<=N; i++) {
    var testx = 14.0 /N * i - 7;
    console.log(testx + ", "+ (lsp.val(testx)));
}

lsp.f.stdout();
